### DISCUSSION TITLE
Set a CRD field immutable

### DISCUSSION THREAD

**Discussion by hjiajing** [kubernetes-sigs/kubebuilder#2829] [Support] [answered]
Hi, 
I met a problem that I have to make a field of a CRD immutable. I tried to use a webhook that uses a `validateUpdate` method to check the old object and the existing object. If the field changes, then the webhook will reject the change. 

I wonder if there is any other way to implement this? Is it possible to implement it with schema ? 

Thanks.

**Comment by camilamacedo86** [comment]
Hi @hjiajing,

If the data is immutable ( users cannot change that by updating the specs on the CR) Why it should be a spec of the CRD in the first place? It seems like that should be a config fixed on the controller/reconciliation.

**Reply by hjiajing** [reply]
Hi Camila, in some cases we need to define a CR. After the CR is created, we cannot change it ( if the users want to change the data, they need to delete and re-create).
But the controller needs some data of the CR to reconcile, so we have to define it in the spec.  Just like some Kubernetes resources, once we create them, some fields cannot be changed unless we delete them and create them again.

**Comment by FillZpp** [comment, marked as answer]
> It seems like that should be a config fixed on the controller/reconciliation.

@camilamacedo86 I think the *immutable* means the field can not be changed once the CR created.

> I wonder if there is any other way to implement this? Is it possible to implement it with schema ?

@hjiajing  No, it seems now CRD does not support immutable fields. You can only validate the update in your webhook.

But I think it's worth to have it in CRD. You can create an issue in k/k repo and see how ppl from sig-api-machinery think about it.

**Reply by hjiajing** [reply]
@FillZpp @camilamacedo86 Thanks for your answer. Maybe I should use a webhook to validate.

**Comment by camilamacedo86** [comment]
Closing since it is answered.
